# Stable Coin Comparison
## Overview

**Stable Coin Yield Comparison** gathers a list of stable coins across various trading platforms and compares their return by simply staking or locking the coins for a specified period of time.  Since these coins are pegged to the US dollar, this method of obtaining yield is considered low risk with minimal effort.  

This analysis is important for setting the benchmark returns on these stable coins.  An arbitrage strategy on the stable coins should at least yield a higher return than the strategies outlined below.  

The data sourced for this was taken from the following platforms:
* [Curve Finance](https://curve.fi/)
* [AAVE](https://aave.com/)
* [Crypto.com](https://crypto.com/us/)
* [Coinbase](https://www.coinbase.com/)

In the [StableOps notebook](./StableOps.ipynb), this notebook statisfies the analysis for the objective of:

* Analysis of low risk yield producing options with stable coins, e.g. staking yields
    * resolve an average low risk yield 

In [1]:
import requests
import json
import pandas as pd
import os
import numpy as np
from utils.utils import epoch_to_datetime
from utils.utils import APY_to_APR

In [2]:
# Create dataframe containing current APR rates for stable coins, source
stable_coin_returns_df = pd.DataFrame(columns=['platform','Coins','staking_method','yield','APR/APY','source'],index=None)

In [3]:
#Hold Column Values for adding addition entries
stable_coin_returns_df_columns=list(stable_coin_returns_df.columns)
stable_coin_returns_df_columns

['platform', 'Coins', 'staking_method', 'yield', 'APR/APY', 'source']

# Curve Finance

In [4]:
# Curve Finance Data
curve_data=['Curve', 'DAI - USDC - USDT','3pool - DAI+USDC+USDT', 0.73, 'APY', 'https://curve.fi/3pool']
curve_dict={}

In [5]:
for i,column in enumerate(stable_coin_returns_df_columns): 
    curve_dict[column]=curve_data[i]

In [6]:
curve_dict

{'platform': 'Curve',
 'Coins': 'DAI - USDC - USDT',
 'staking_method': '3pool - DAI+USDC+USDT',
 'yield': 0.73,
 'APR/APY': 'APY',
 'source': 'https://curve.fi/3pool'}

In [7]:
#Add Curve to DataFrame:
stable_coin_returns_df.loc[0]=curve_dict
stable_coin_returns_df

platform              Coins         staking_method  yield APR/APY  \
0    Curve  DAI - USDC - USDT  3pool - DAI+USDC+USDT   0.73     APY   

                   source  
0  https://curve.fi/3pool

# AAVE

In [8]:
# AAVE Stablecoin Data:
aave_data_BUSD=['AAVE','BUSD', 'Supply BUSD', 2.17, 'APY', 'https://app.aave.com/']
aave_data_DAI=['AAVE','DAI', 'Supply DAI', 2.23, 'APY', 'https://app.aave.com/']
aave_data_GUSD=['AAVE','GUSD', 'Supply GUSD', 1.66, 'APY', 'https://app.aave.com/']
aave_data_sUSD=['AAVE','sUSD', 'Supply sUSD', 2.09, 'APY', 'https://app.aave.com/']
aave_data_TUSD=['AAVE','TUSD', 'Supply TUSD', 1.94, 'APY', 'https://app.aave.com/']
aave_data_USDC=['AAVE','USDC', 'Supply USDC', 1.81, 'APY', 'https://app.aave.com/']
aave_data_USDP=['AAVE','USDP', 'Supply USDP', 2.74, 'APY', 'https://app.aave.com/']
aave_data_USDT=['AAVE','USDT', 'Supply USDT', 1.90, 'APY', 'https://app.aave.com/']
aave_data_UST=['AAVE','UST', 'Supply UST', 48.09, 'APY', 'https://app.aave.com/']

In [9]:
# AAVE to df:
AAVE_data= [aave_data_BUSD,
 aave_data_DAI,
 aave_data_GUSD,
 aave_data_sUSD,
 aave_data_TUSD,
 aave_data_USDC,
 aave_data_USDP,
 aave_data_USDT,
 aave_data_UST]

In [10]:
AAVE_dict={}
for j,entry in enumerate(AAVE_data):
    AAVE_dict[j+1]={}
    for i,column in enumerate(stable_coin_returns_df_columns): 
        AAVE_dict[j+1][column]=entry[i]

In [11]:
AAVE_dict

{1: {'platform': 'AAVE',
  'Coins': 'BUSD',
  'staking_method': 'Supply BUSD',
  'yield': 2.17,
  'APR/APY': 'APY',
  'source': 'https://app.aave.com/'},
 2: {'platform': 'AAVE',
  'Coins': 'DAI',
  'staking_method': 'Supply DAI',
  'yield': 2.23,
  'APR/APY': 'APY',
  'source': 'https://app.aave.com/'},
 3: {'platform': 'AAVE',
  'Coins': 'GUSD',
  'staking_method': 'Supply GUSD',
  'yield': 1.66,
  'APR/APY': 'APY',
  'source': 'https://app.aave.com/'},
 4: {'platform': 'AAVE',
  'Coins': 'sUSD',
  'staking_method': 'Supply sUSD',
  'yield': 2.09,
  'APR/APY': 'APY',
  'source': 'https://app.aave.com/'},
 5: {'platform': 'AAVE',
  'Coins': 'TUSD',
  'staking_method': 'Supply TUSD',
  'yield': 1.94,
  'APR/APY': 'APY',
  'source': 'https://app.aave.com/'},
 6: {'platform': 'AAVE',
  'Coins': 'USDC',
  'staking_method': 'Supply USDC',
  'yield': 1.81,
  'APR/APY': 'APY',
  'source': 'https://app.aave.com/'},
 7: {'platform': 'AAVE',
  'Coins': 'USDP',
  'staking_method': 'Supply USDP',

In [12]:
for key in AAVE_dict:
    stable_coin_returns_df.loc[key]=AAVE_dict[key]

In [13]:
stable_coin_returns_df

platform              Coins         staking_method  yield APR/APY  \
0    Curve  DAI - USDC - USDT  3pool - DAI+USDC+USDT   0.73     APY   
1     AAVE               BUSD            Supply BUSD   2.17     APY   
2     AAVE                DAI             Supply DAI   2.23     APY   
3     AAVE               GUSD            Supply GUSD   1.66     APY   
4     AAVE               sUSD            Supply sUSD   2.09     APY   
5     AAVE               TUSD            Supply TUSD   1.94     APY   
6     AAVE               USDC            Supply USDC   1.81     APY   
7     AAVE               USDP            Supply USDP   2.74     APY   
8     AAVE               USDT            Supply USDT   1.90     APY   
9     AAVE                UST             Supply UST  48.09     APY   

                   source  
0  https://curve.fi/3pool  
1   https://app.aave.com/  
2   https://app.aave.com/  
3   https://app.aave.com/  
4   https://app.aave.com/  
5   https://app.aave.com/  
6   https://app.aave.com/  
7   https://app.aave.com/  
8   https://app.aave.com/  
9   https://app.aave.com/

# Crypto.com

In [14]:
# Crypto.com Stablecoin Data:
cro_data_USDT=['crypto.com','USDT', 'Stake USDT', 6.00, 'APR', 'https://crypto.com/us/earn']
cro_data_USDC=['crypto.com','USDC', 'Stake USDC', 6.00, 'APR', 'https://crypto.com/us/earn']
cro_data_DAI=['crypto.com','DAI', 'Stake DAI', 6.00, 'APR', 'https://crypto.com/us/earn']
cro_data_USDP=['crypto.com','USDP', 'Stake USDP', 6.00, 'APR', 'https://crypto.com/us/earn']
cro_data_TUSD=['crypto.com','TUSD', 'Stake TUSD', 6.00, 'APR', 'https://crypto.com/us/earn']

In [15]:
cro_data=[cro_data_USDT,
          cro_data_USDC,
          cro_data_DAI,
          cro_data_USDP,
          cro_data_TUSD]

In [16]:
cro_dict={}
for j,entry in enumerate(cro_data):
    cro_dict[j+10]={}
    for i,column in enumerate(stable_coin_returns_df_columns): 
        cro_dict[j+10][column]=entry[i]

In [17]:
for key in cro_dict:
    stable_coin_returns_df.loc[key]=cro_dict[key]

In [18]:
stable_coin_returns_df

platform              Coins         staking_method  yield APR/APY  \
0        Curve  DAI - USDC - USDT  3pool - DAI+USDC+USDT   0.73     APY   
1         AAVE               BUSD            Supply BUSD   2.17     APY   
2         AAVE                DAI             Supply DAI   2.23     APY   
3         AAVE               GUSD            Supply GUSD   1.66     APY   
4         AAVE               sUSD            Supply sUSD   2.09     APY   
5         AAVE               TUSD            Supply TUSD   1.94     APY   
6         AAVE               USDC            Supply USDC   1.81     APY   
7         AAVE               USDP            Supply USDP   2.74     APY   
8         AAVE               USDT            Supply USDT   1.90     APY   
9         AAVE                UST             Supply UST  48.09     APY   
10  crypto.com               USDT             Stake USDT   6.00     APR   
11  crypto.com               USDC             Stake USDC   6.00     APR   
12  crypto.com                DAI              Stake DAI   6.00     APR   
13  crypto.com               USDP             Stake USDP   6.00     APR   
14  crypto.com               TUSD             Stake TUSD   6.00     APR   

                        source  
0       https://curve.fi/3pool  
1        https://app.aave.com/  
2        https://app.aave.com/  
3        https://app.aave.com/  
4        https://app.aave.com/  
5        https://app.aave.com/  
6        https://app.aave.com/  
7        https://app.aave.com/  
8        https://app.aave.com/  
9        https://app.aave.com/  
10  https://crypto.com/us/earn  
11  https://crypto.com/us/earn  
12  https://crypto.com/us/earn  
13  https://crypto.com/us/earn  
14  https://crypto.com/us/earn

# Coinbase

In [19]:
# coinbase.com Stablecoin Data:
coinbase_data_DAI=['coinbase.com','DAI', 'Stake DAI', 0.15, 'APY', 'https://www.coinbase.com/staking']


In [20]:
stable_coin_returns_df.loc[15]=coinbase_data_DAI

# Full Dataframe & Results

In [21]:
stable_coin_returns_df

platform              Coins         staking_method  yield APR/APY  \
0          Curve  DAI - USDC - USDT  3pool - DAI+USDC+USDT   0.73     APY   
1           AAVE               BUSD            Supply BUSD   2.17     APY   
2           AAVE                DAI             Supply DAI   2.23     APY   
3           AAVE               GUSD            Supply GUSD   1.66     APY   
4           AAVE               sUSD            Supply sUSD   2.09     APY   
5           AAVE               TUSD            Supply TUSD   1.94     APY   
6           AAVE               USDC            Supply USDC   1.81     APY   
7           AAVE               USDP            Supply USDP   2.74     APY   
8           AAVE               USDT            Supply USDT   1.90     APY   
9           AAVE                UST             Supply UST  48.09     APY   
10    crypto.com               USDT             Stake USDT   6.00     APR   
11    crypto.com               USDC             Stake USDC   6.00     APR   
12    crypto.com                DAI              Stake DAI   6.00     APR   
13    crypto.com               USDP             Stake USDP   6.00     APR   
14    crypto.com               TUSD             Stake TUSD   6.00     APR   
15  coinbase.com                DAI              Stake DAI   0.15     APY   

                              source  
0             https://curve.fi/3pool  
1              https://app.aave.com/  
2              https://app.aave.com/  
3              https://app.aave.com/  
4              https://app.aave.com/  
5              https://app.aave.com/  
6              https://app.aave.com/  
7              https://app.aave.com/  
8              https://app.aave.com/  
9              https://app.aave.com/  
10        https://crypto.com/us/earn  
11        https://crypto.com/us/earn  
12        https://crypto.com/us/earn  
13        https://crypto.com/us/earn  
14        https://crypto.com/us/earn  
15  https://www.coinbase.com/staking

## Datacleaning

* Since many of the yields quoted on the various platforms are quoted in annual percentage yield (APY), the return percentages need to be converted to an annual percentage rate (APR).  As described by the platform, the APY is the expected return after a year of compounding at the current rate.  To make a fair assessment, APY needs to be converted to an APR.

* To streamline this process, the function ```APY_to_APR(apy,N)``` is written to help with the conversion on all of the datacolumns, and can be found in the [utils directory](./utils/APY_to_APR.py).  It takes two arguments, the APY to convert ```apy``` and the number of compounding periods in a year ```N```.  The compounding period on most platforms is on a weekly basis or $N = 52$.  The function then returns a corresponding APR.

In [22]:
# Example calculation using APY_to_APR(apy,N)
APY_to_APR(0.06,52)

0.058301567106710195

In [23]:
# Using APY_to_APR, an APR column is created in stable_coin_returns_df by passing it through a lambda function on all elements.
stable_coin_returns_df['APR'] = stable_coin_returns_df['yield'].apply(lambda x: APY_to_APR(x,52)) / 100

In [24]:
stable_coin_returns_df

platform              Coins         staking_method  yield APR/APY  \
0          Curve  DAI - USDC - USDT  3pool - DAI+USDC+USDT   0.73     APY   
1           AAVE               BUSD            Supply BUSD   2.17     APY   
2           AAVE                DAI             Supply DAI   2.23     APY   
3           AAVE               GUSD            Supply GUSD   1.66     APY   
4           AAVE               sUSD            Supply sUSD   2.09     APY   
5           AAVE               TUSD            Supply TUSD   1.94     APY   
6           AAVE               USDC            Supply USDC   1.81     APY   
7           AAVE               USDP            Supply USDP   2.74     APY   
8           AAVE               USDT            Supply USDT   1.90     APY   
9           AAVE                UST             Supply UST  48.09     APY   
10    crypto.com               USDT             Stake USDT   6.00     APR   
11    crypto.com               USDC             Stake USDC   6.00     APR   
12    crypto.com                DAI              Stake DAI   6.00     APR   
13    crypto.com               USDP             Stake USDP   6.00     APR   
14    crypto.com               TUSD             Stake TUSD   6.00     APR   
15  coinbase.com                DAI              Stake DAI   0.15     APY   

                              source       APR  
0             https://curve.fi/3pool  0.005510  
1              https://app.aave.com/  0.011666  
2              https://app.aave.com/  0.011858  
3              https://app.aave.com/  0.009876  
4              https://app.aave.com/  0.011405  
5              https://app.aave.com/  0.010897  
6              https://app.aave.com/  0.010435  
7              https://app.aave.com/  0.013360  
8              https://app.aave.com/  0.010757  
9              https://app.aave.com/  0.040431  
10        https://crypto.com/us/earn  0.019828  
11        https://crypto.com/us/earn  0.019828  
12        https://crypto.com/us/earn  0.019828  
13        https://crypto.com/us/earn  0.019828  
14        https://crypto.com/us/earn  0.019828  
15  https://www.coinbase.com/staking  0.001399

In [25]:
# The yields initially designated as an APR value are copied over to the APR column
stable_coin_returns_df.loc[(stable_coin_returns_df['APR/APY']=='APR'),'APR'] = stable_coin_returns_df['yield']
stable_coin_returns_df

platform              Coins         staking_method  yield APR/APY  \
0          Curve  DAI - USDC - USDT  3pool - DAI+USDC+USDT   0.73     APY   
1           AAVE               BUSD            Supply BUSD   2.17     APY   
2           AAVE                DAI             Supply DAI   2.23     APY   
3           AAVE               GUSD            Supply GUSD   1.66     APY   
4           AAVE               sUSD            Supply sUSD   2.09     APY   
5           AAVE               TUSD            Supply TUSD   1.94     APY   
6           AAVE               USDC            Supply USDC   1.81     APY   
7           AAVE               USDP            Supply USDP   2.74     APY   
8           AAVE               USDT            Supply USDT   1.90     APY   
9           AAVE                UST             Supply UST  48.09     APY   
10    crypto.com               USDT             Stake USDT   6.00     APR   
11    crypto.com               USDC             Stake USDC   6.00     APR   
12    crypto.com                DAI              Stake DAI   6.00     APR   
13    crypto.com               USDP             Stake USDP   6.00     APR   
14    crypto.com               TUSD             Stake TUSD   6.00     APR   
15  coinbase.com                DAI              Stake DAI   0.15     APY   

                              source       APR  
0             https://curve.fi/3pool  0.005510  
1              https://app.aave.com/  0.011666  
2              https://app.aave.com/  0.011858  
3              https://app.aave.com/  0.009876  
4              https://app.aave.com/  0.011405  
5              https://app.aave.com/  0.010897  
6              https://app.aave.com/  0.010435  
7              https://app.aave.com/  0.013360  
8              https://app.aave.com/  0.010757  
9              https://app.aave.com/  0.040431  
10        https://crypto.com/us/earn  6.000000  
11        https://crypto.com/us/earn  6.000000  
12        https://crypto.com/us/earn  6.000000  
13        https://crypto.com/us/earn  6.000000  
14        https://crypto.com/us/earn  6.000000  
15  https://www.coinbase.com/staking  0.001399

In [26]:
# By using the summary statistics on the dataframe, we can access the current APR for stable coins using yield.
stable_coin_returns_df.describe()

yield        APR
count  16.000000  16.000000
mean    5.969375   1.883600
std    11.425210   2.866305
min     0.150000   0.001399
25%     1.877500   0.010676
50%     2.200000   0.011762
75%     6.000000   6.000000
max    48.090000   6.000000

* Therefore a reasonable yield to expect by staking a stable coin is **2.73 %**.  

### Yield by Platform

In [27]:
average_return_by_coin=stable_coin_returns_df.groupby('platform')
average_return_by_coin.describe()

yield                                                       APR  \
             count      mean        std   min   25%   50%   75%    max count   
platform                                                                       
AAVE           9.0  7.181111  15.343955  1.66  1.90  2.09  2.23  48.09   9.0   
Curve          1.0  0.730000        NaN  0.73  0.73  0.73  0.73   0.73   1.0   
coinbase.com   1.0  0.150000        NaN  0.15  0.15  0.15  0.15   0.15   1.0   
crypto.com     5.0  6.000000   0.000000  6.00  6.00  6.00  6.00   6.00   5.0   

                                                                          \
                  mean       std       min       25%       50%       75%   
platform                                                                   
AAVE          0.014521  0.009767  0.009876  0.010757  0.011405  0.011858   
Curve         0.005510       NaN  0.005510  0.005510  0.005510  0.005510   
coinbase.com  0.001399       NaN  0.001399  0.001399  0.001399  0.001399   
crypto.com    6.000000  0.000000  6.000000  6.000000  6.000000  6.000000   

                        
                   max  
platform                
AAVE          0.040431  
Curve         0.005510  
coinbase.com  0.001399  
crypto.com    6.000000

### Yield by Coin

In [28]:
average_return_by_coin=stable_coin_returns_df.sort_values(by='APR').groupby('Coins')
average_return_by_coin.describe()

yield                                                        \
                  count       mean       std    min      25%     50%      75%   
Coins                                                                           
BUSD                1.0   2.170000       NaN   2.17   2.1700   2.170   2.1700   
DAI                 3.0   2.793333  2.965406   0.15   1.1900   2.230   4.1150   
DAI - USDC - USDT   1.0   0.730000       NaN   0.73   0.7300   0.730   0.7300   
GUSD                1.0   1.660000       NaN   1.66   1.6600   1.660   1.6600   
TUSD                2.0   3.970000  2.870854   1.94   2.9550   3.970   4.9850   
USDC                2.0   3.905000  2.962777   1.81   2.8575   3.905   4.9525   
USDP                2.0   4.370000  2.305168   2.74   3.5550   4.370   5.1850   
USDT                2.0   3.950000  2.899138   1.90   2.9250   3.950   4.9750   
UST                 1.0  48.090000       NaN  48.09  48.0900  48.090  48.0900   
sUSD                1.0   2.090000       NaN   2.09   2.0900   2.090   2.0900   

                           APR                                          \
                     max count      mean       std       min       25%   
Coins                                                                    
BUSD                2.17   1.0  0.011666       NaN  0.011666  0.011666   
DAI                 6.00   3.0  2.004419  3.460278  0.001399  0.006629   
DAI - USDC - USDT   0.73   1.0  0.005510       NaN  0.005510  0.005510   
GUSD                1.66   1.0  0.009876       NaN  0.009876  0.009876   
TUSD                6.00   2.0  3.005448  4.234936  0.010897  1.508173   
USDC                6.00   2.0  3.005218  4.235262  0.010435  1.507826   
USDP                6.00   2.0  3.006680  4.233194  0.013360  1.510020   
USDT                6.00   2.0  3.005378  4.235034  0.010757  1.508068   
UST                48.09   1.0  0.040431       NaN  0.040431  0.040431   
sUSD                2.09   1.0  0.011405       NaN  0.011405  0.011405   

                                                 
                        50%       75%       max  
Coins                                            
BUSD               0.011666  0.011666  0.011666  
DAI                0.011858  3.005929  6.000000  
DAI - USDC - USDT  0.005510  0.005510  0.005510  
GUSD               0.009876  0.009876  0.009876  
TUSD               3.005448  4.502724  6.000000  
USDC               3.005218  4.502609  6.000000  
USDP               3.006680  4.503340  6.000000  
USDT               3.005378  4.502689  6.000000  
UST                0.040431  0.040431  0.040431  
sUSD               0.011405  0.011405  0.011405

## Conclusions from Analysis

* While some of the returns from the stable coin yield farming look attractive, we have to be careful in choosing a benchmark yield for our analysis, i.e. one that will not change significantly over time.  Upon closer inspection, we decided to disregard the following from our analysis:

    - The stable coin yield of 6.00 % offered by crypto.com is a promotional rate and also requires the purchase of the crypto coin CRO to activate.  So the returns quoted by crypto.com will not be included.
    
    
    - The stable coin UST destabilized in recent dates (after this analysis was initially conducted on May 7, 2022), and its outlier yield of 48.1 % will not be included in our statistics.
    
    
    - The platforms of coinbase.com and Curve offer yields well below that of AAVE, but have comparable risk.  Therefore, coinbase.com and Curve will be disregarded in calculating our benchmark return.
    


* Based on this dataset, we conclude that AAVE is the best platform for representing the low risk yield for stable coins.  To generate a reasonable expected return on stable coins, the returns of each stablecoin (except UST) on AAVE are averaged together and an annual return rate is calculated.  

In [29]:
# Dropping extreme outlier UST from average_return_by_coin
#average_return_by_coin_AAVE = average_return_by_coin.loc[average_return_by_coin['platform']=='AAVE'

stable_coin_returns_AAVE_df = stable_coin_returns_df.loc[stable_coin_returns_df['platform']=='AAVE']

In [30]:
stable_coin_returns_AAVE_df= stable_coin_returns_AAVE_df.loc[stable_coin_returns_df['Coins']!='UST']
stable_coin_returns_AAVE_df

platform Coins staking_method  yield APR/APY                 source  \
1     AAVE  BUSD    Supply BUSD   2.17     APY  https://app.aave.com/   
2     AAVE   DAI     Supply DAI   2.23     APY  https://app.aave.com/   
3     AAVE  GUSD    Supply GUSD   1.66     APY  https://app.aave.com/   
4     AAVE  sUSD    Supply sUSD   2.09     APY  https://app.aave.com/   
5     AAVE  TUSD    Supply TUSD   1.94     APY  https://app.aave.com/   
6     AAVE  USDC    Supply USDC   1.81     APY  https://app.aave.com/   
7     AAVE  USDP    Supply USDP   2.74     APY  https://app.aave.com/   
8     AAVE  USDT    Supply USDT   1.90     APY  https://app.aave.com/   

        APR  
1  0.011666  
2  0.011858  
3  0.009876  
4  0.011405  
5  0.010897  
6  0.010435  
7  0.013360  
8  0.010757

In [31]:
#Descriptive stats
stable_coin_returns_AAVE_df.describe()

yield       APR
count  8.000000  8.000000
mean   2.067500  0.011282
std    0.330875  0.001064
min    1.660000  0.009876
25%    1.877500  0.010676
50%    2.015000  0.011151
75%    2.185000  0.011714
max    2.740000  0.013360

* Based on the summary statistics below, the most representative, low risk yield to expect from **stable coin yields is 1.13 % with a standard deviation of 0.11 %**. 